src/modal/modal-window.ts

```diff
@@ -52,40 +56,51 @@ export class NgbModalWindow implements OnInit,
   @Output('dismiss') dismissEvent = new EventEmitter();
 
   constructor(
-      @Inject(DOCUMENT) private _document: any, private _elRef: ElementRef<HTMLElement>, private _zone: NgZone) {
-    _zone.runOutsideAngular(() => {
-      fromEvent<KeyboardEvent>(this._elRef.nativeElement, 'keydown')
+      @Inject(DOCUMENT) private _document: any, private _elRef: ElementRef<HTMLElement>, private _zone: NgZone) {}
+
+  dismiss(reason): void { this.dismissEvent.emit(reason); }
+
+  ngOnInit() { this._elWithFocus = this._document.activeElement; }
+
+  ngAfterViewInit() {
+    const {nativeElement} = this._elRef;
+    this._zone.runOutsideAngular(() => {
+
+      fromEvent<KeyboardEvent>(nativeElement, 'keydown')
           .pipe(
               takeUntil(this.dismissEvent),
               // tslint:disable-next-line:deprecation
               filter(e => e.which === Key.Escape && this.keyboard))
           .subscribe(event => requestAnimationFrame(() => {
                        if (!event.defaultPrevented) {
-                         _zone.run(() => this.dismiss(ModalDismissReasons.ESC));
+                         this._zone.run(() => this.dismiss(ModalDismissReasons.ESC));
                        }
                      }));
 
-      const mouseDowns$ = fromEvent<MouseEvent>(this._elRef.nativeElement, 'mousedown')
-                              .pipe(
-                                  takeUntil(this.dismissEvent),
-                                  map(e => this.backdrop === true && this._elRef.nativeElement === e.target));
+      // This is done to prevent modal from closing when pressing the mouse inside the dialog and releasing it outside
+      fromEvent<MouseEvent>(this._dialogEl.nativeElement, 'mousedown')
+          .pipe(takeUntil(this.dismissEvent), tap(() => {
+                  this._preventClose = false;
+                  fromEvent<MouseEvent>(nativeElement, 'mouseup')
+                      .pipe(takeUntil(this.dismissEvent), take(1))
+                      .subscribe(({target}) => { this._preventClose = nativeElement === target; });
+                }))
+          .subscribe();
 
-      fromEvent<MouseEvent>(this._elRef.nativeElement, 'mouseup')
-          .pipe(takeUntil(this.dismissEvent), withLatestFrom(mouseDowns$), filter(([_, shouldClose]) => shouldClose))
-          .subscribe(() => this._zone.run(() => this.dismiss(ModalDismissReasons.BACKDROP_CLICK)));
+      // We have to use 'click' event to prevent modal from closing when the modal is too big and we have a scrollbar
```

Comment:
I think current wording can be misleading, as the 'click' is the primary way to dismis modal via backdrop, mosue down/up is used to prevent accidental close when released from 'down > drag > release' action when fired 'click' closes dialog.

Reply:
Thanks, apparently was thinking about the issue when writing this :) Will update comments